utils/modules/edge_utils.py

```diff
@@ -62,8 +90,29 @@ def adjacent_faces(edges, pols):
                 ad_faces[idx] += 1
     return ad_faces
 
+def adjacent_faces(edges, pols):
+    '''calculate adjacent faces '''
+    e_sorted = [sorted(e) for e in edges]
+    ad_faces = [[] for e in edges]
+    for pol in pols:
+        for edge in zip(pol, pol[1:] + [pol[0]]):
+            e_s = sorted(edge)
+            if e_s in e_sorted:
+                idx = e_sorted.index(e_s)
+                ad_faces[idx] += [pol]
+    return ad_faces
+def faces_angle_full(vertices, edges, faces):
+    '''angle between faces of each edge (only first two faces)'''
+    bm = bmesh_from_pydata(vertices, edges, faces, normal_update=True)
+    normals = [tuple(face.normal) for face in bm.faces]
+    bm.free()
+    vals = faces_angle(normals, edges, faces)
+    return vals
```

Comment:
There could be some cases in which it would be nice to have the `function(bm)` but not for all of them, in this case the function is also using original edges (that get messed in bmesh) and faces so it would need to be `function(bm, edges, faces)`....
The efficiency of the function depends on the use, for the Component Analyzer that just calls one of this function per node it was better to perform the conversion inside the function if it was needed.
I think that if any developer finds more convenient function(bm) they can make a new function taking the example of the previous one

Reply:
The less Sverchok has functions the better. I would probably like to use some of this functions in other nodes but your approach does not let it. I does not see reasons why all this functions are in `utils` folder if they can be efficiently used only by `Component Analyzer` node. Or probably names of folders should be less ambitious, the name can be like `Component Analyzer node utils` so everyone would know that the fuctions are mostly only for the node usage.

Reply:
Right now I think is better to have more smaller functions and a bigger one with many conditional cases inside, but i guess there is not a truth for all the cases... this node is a prof of concept and could be expanded or reduced i think that is not a bad approach to a muti-function node work and open the mind of the user
The functions can be used efficiently in many other cases, obviously not all of them, they could be called from a script node (without knowing anything about converting to bmesh) or get inside other algorithms. I started to fill the vertex_utils, edges_utils, pols_utils and matrix_utils to light the node code and in case someone can use them too, if you find it disturbing i can change the filenames and folder

Reply:
Right now I think is better to have more smaller functions and a bigger one with many conditional cases inside, but i guess there is not a truth for all the cases... this node is a prof of concept and could be expanded or reduced i think that is not a bad approach to a muti-function node work and open the mind of the user
The functions can be used efficiently in many other cases, obviously not all of them, they could be called from a script node (without knowing anything about converting to bmesh) or get inside other algorithms. I started to fill the vertex_utils, edges_utils, pols_utils and matrix_utils to light the node code and in case someone can use them too, if you find it disturbing i can change the filenames and folder